# Pandas DataFrames 101 - Learning Notebook

----

<img src="https://miro.medium.com/v2/resize:fit:1400/1*Q55X1gfHjKi1knx96UVdZQ.png" alt="Description" width="450">

## Contents

In this notebook we will cover the following:

1. **Introduction to DataFrames**
    1. What is a DataFrame?
    2. Reading data into a dataframe
2. **Previewing and describing a DataFrame**
    1. Common methods for a quick overview
3. **Selecting and filtering data**
    1. Selecting rows and columns
    2. Using masks
4. **Analyzing and aggregating data**
    1. Analysing numerical data
    2. Analyzing categorical data
    3. Using `.group_by()`

## Imports

In [1]:
import pandas as pd

# 1. Introduction to DataFrames

## Reading data into a DataFrame

In [2]:
# isto lê uma base de dados que está guardada na internet e cria um DataFrame a partir dela.
df = pd.read_csv('https://raw.githubusercontent.com/Hospital-Da-Luz-Learning-Health/MLCatolica24/main/Aula%203%20-%20Data%20Analysis%20%26%20Statistics/Pandas%20DataFrames%20101/data/heart_disease_cleaned.csv').set_index('id')

# 2. Previewing and describing a DataFrame

## Common methods for a quick overview

Let's take a sneak peak at this data:

In [3]:
df.head() # apresenta as primeiras 5 linhas do DataFrame

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
192,43,male,asymptomatic,132,247,1,left ventricular hypertrophy,143,1,0.1,fat,NaN,reversible defect,>50%
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%
0,63,male,typical angina,145,233,1,left ventricular hypertrophy,150,0,2.3,downsloping,0.0,fixed defect,<50%
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
284,61,male,asymptomatic,148,203,0,normal,161,0,0.0,upsloping,1.0,reversible defect,>50%


In [4]:
df.head(3) # vocês podem definir quantas querem ver, por exemplo 3

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
192,43,male,asymptomatic,132,247,1,left ventricular hypertrophy,143,1,0.1,fat,NaN,reversible defect,>50%
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%
0,63,male,typical angina,145,233,1,left ventricular hypertrophy,150,0,2.3,downsloping,0.0,fixed defect,<50%


In [5]:
df.shape # -> (número de linhas, número de colunas) do DataFrames

(303, 14)

# Accessing Data

In [12]:
df.columns # devolve uma lista com o nome de todas as colunas do DataFrame

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'Diagnosed'],
      dtype='object')

In [14]:
df.index # devolve uma lista com todos os índices do DataFrame

Index([192, 189,   0,  23, 284, 177,  89, 290, 127,  56,
       ...
       163, 169, 186, 244, 139,   9, 109, 185, 111,  36],
      dtype='int64', name='id', length=303)

In [11]:
df['age'] # devolve todos os valores da coluna "idade"

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

We selected just the `age` but the output shows two columns of data:
1. The second column contains the actual `age` data (43,69,63, etc.)
2. While the first column contains the index of each row (which patient it refers to) (192,189,0, etc.)

In [23]:
df['weight'] # Se tentam aceder a uma coluna que não existe dá este tipo de erro (para futura referência)

KeyError: 'weight'

----

You can select more than one column! Just pass a list of column names

In [20]:
columns_i_want_to_select = ["age", "slope"]

df[columns_i_want_to_select]

,age,slope
id,,
192,43,fat
189,69,fat
0,63,downsloping
23,58,upsloping
284,61,upsloping
...,...,...
9,53,downsloping
109,39,fat
185,63,upsloping


In [22]:
# reparem que o que está acima é equivalente a fazer:
df[["age","slope"]]
# daí ser necessário usar dois pares de parêntesis quando querem selecionar mais do que uma coluna.

,age,slope
id,,
192,43,fat
189,69,fat
0,63,downsloping
23,58,upsloping
284,61,upsloping
...,...,...
9,53,downsloping
109,39,fat
185,63,upsloping


# Selecting data (loc and iloc)

## .loc

`loc` is short for locate

In [ ]:
df.loc[:, :] # isto seleciona todas as linhas e colunas (redundante, sim)

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
192,43,male,asymptomatic,132,247,1,left ventricular hypertrophy,143,1,0.1,fat,NaN,reversible defect,>50%
189,69,male,non-anginal pain,140,254,0,left ventricular hypertrophy,146,0,2.0,fat,3.0,reversible defect,>50%
0,63,male,typical angina,145,233,1,left ventricular hypertrophy,150,0,2.3,downsloping,0.0,fixed defect,<50%
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
284,61,male,asymptomatic,148,203,0,normal,161,0,0.0,upsloping,1.0,reversible defect,>50%
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
9,53,male,asymptomatic,140,203,1,left ventricular hypertrophy,155,1,3.1,downsloping,0.0,reversible defect,>50%
109,39,male,asymptomatic,118,219,0,normal,140,0,1.2,fat,0.0,reversible defect,>50%
185,63,female,atypical angina,140,195,0,normal,179,0,0.0,upsloping,2.0,normal,<50%


The `...` strike again in the output above. Again, it's just a display optimization, the 303 rows are still there.

In [ ]:
df.loc[:, 'age'] # devolve todos os valores da coluna "age", é o mesmo que fazer df['age']

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

In [24]:
df.loc[:,['age','slope']] # devolve todos os valores das colunas "age" e "slope", é o mesmo que fazer df[['age','slope']]

,age,slope
id,,
192,43,fat
189,69,fat
0,63,downsloping
23,58,upsloping
284,61,upsloping
...,...,...
9,53,downsloping
109,39,fat
185,63,upsloping


In [26]:
df.loc[23,:] # devolve toda a informação do paciente 23

age                                    58
sex                                  male
cp                       non-anginal pain
trestbps                              132
chol                                  224
fbs                                     0
restecg      left ventricular hypertrophy
thalach                               173
exang                                   0
oldpeak                               3.2
slope                           upsloping
ca                                    2.0
thal                    reversible defect
Diagnosed                            >50%
Name: 23, dtype: object

In [27]:
df.loc[23, 'age'] # devolve a idade do paciente 23

58

We can create programs around this data:

In [29]:
if df.loc[23, 'age'] > 50:
    print("O paciente com id 23 tem mais de 50 anos.")

O paciente com id 23 tem mais de 50 anos.


We can select multiple patients

In [30]:
my_patients = [23, 185]

df.loc[my_patients, :] # devolve toda a informação dos pacientes 23 e 185

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
185,63,female,atypical angina,140,195,0,normal,179,0,0.0,upsloping,2.0,normal,<50%


The above is the same as doing:

In [31]:
df.loc[[23, 185], :]

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,Diagnosed
id,,,,,,,,,,,,,,
23,58,male,non-anginal pain,132,224,0,left ventricular hypertrophy,173,0,3.2,upsloping,2.0,reversible defect,>50%
185,63,female,atypical angina,140,195,0,normal,179,0,0.0,upsloping,2.0,normal,<50%


And to select multiple patients and multiple columns:

In [33]:
df.loc[[44, 21], ['restecg', "ca"]]

,restecg,ca
id,,
44,left ventricular hypertrophy,0.0
21,left ventricular hypertrophy,0.0


## .iloc

With iloc you select by the index position of the rows and/or column and not their values.

In [6]:
df.iloc[0,:] # devolve toda a informação do doente na primeira posição da dataframe

age                                    43
sex                                  male
cp                           asymptomatic
trestbps                              132
chol                                  247
fbs                                     1
restecg      left ventricular hypertrophy
thalach                               143
exang                                   1
oldpeak                               0.1
slope                                 fat
ca                                    NaN
thal                    reversible defect
Diagnosed                            >50%
Name: 192, dtype: object

In [7]:
df.iloc[:,0] # devolve a primeira coluna do DataFrame de todos os doentes

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

In [8]:
df.iloc[[0,10], [0,1]] # devolve a primeira e segunda coluna dos doentes na primeira e décima posição do DataFrame

,age,sex
id,,
192,43,male
96,59,male


## Selection rows and columns

In [6]:
df['age'] # selects a column and return a pandas series

id
192    43
189    69
0      63
23     58
284    61
       ..
9      53
109    39
185    63
111    56
36     43
Name: age, Length: 303, dtype: int64

In [7]:
df.loc[:189,'age'] # does the same thing

id
192    43
189    69
Name: age, dtype: int64

We can access the index like so:

In [8]:
df.index

Index([192, 189,   0,  23, 284, 177,  89, 290, 127,  56,
       ...
       163, 169, 186, 244, 139,   9, 109, 185, 111,  36],
      dtype='int64', name='id', length=303)

And this works like a list, so let's see who is the patient in the 41st line

In [9]:
df.index[40]

235

> What is the `age` of the patient whose `id` is `235`?

> What is the `slope` and `cp` of the patient whose `id` is `177`?

> What is the `sex` and `thal` of patients `23`, `199` and `102`?